### ISSUE TITLE
How to export parametric shapes to openSCAD？

### ISSUE THREAD

**Issue by MasterOrigamii** [kimren227/CSGStumpNet#11]
I didn't find the code to export parametric shapes to **openSCAD**, could you please point it out or provide it?
Thank you so much!

**Comment by kimren227** [comment]
This should get you started. Hope it helps.

```
from scipy.spatial.transform import Rotation as R
import numpy as np
from openpyscad import *
from pyquaternion import Quaternion
import time

start_time = time.time()
np.set_printoptions(suppress=True)

data = np.load("63.npy", allow_pickle=True)

primitives = data.item()['primitive'].transpose()
boxes = primitives[:,:10]
cylinders = primitives[:,10:18]
spheres = primitives[:,18:26]
cones = primitives[:,26:]

cvx = data.item()['cvx']
ccv = data.item()['ccv']
cad_cubes = []
cad_spheres = []
cad_cylinders = []
cad_cones = []

for box in boxes:
    quaternion = Quaternion(box[:4]) #[w,x,y,z]
    inverse = quaternion.inverse
    quaternion = np.asarray([inverse[1], inverse[2], inverse[3],inverse[0]]) # [x,y,z,w]
    r = R.from_quat(quaternion).as_euler('xyz', degrees=True)

    translation = box[4:7]
    dim = abs(box[7:])*2
    cad_cubes.append(Cube(dim.tolist(),center=True).rotate([r[0], r[1], r[2]]).translate([translation[0], translation[1], translation[2]]))

for cylinder in cylinders:
    quaternion = Quaternion(cylinder[:4]) #[w,x,y,z]
    inverse = quaternion.inverse
    quaternion = np.asarray([inverse[1], inverse[2], inverse[3],inverse[0]])
    r = R.from_quat(quaternion).as_euler('xyz', degrees=True)

    translation = cylinder[4:7]
    radius = abs(cylinder[-1])
    cad_cylinders.append(Cylinder(h=10, r=radius, _fn=100).translate([0,0,-5]).rotate([r[0], r[1], r[2]]).translate([translation[0], translation[1], translation[2]]))

for sphere in spheres:
    quaternion = Quaternion(sphere[:4]) #[w,x,y,z]
    inverse = quaternion.inverse
    quaternion = np.asarray([inverse[1], inverse[2], inverse[3],inverse[0]])
    r = R.from_quat(quaternion).as_euler('xyz', degrees=True)

    translation = sphere[4:7]
    radius = abs(sphere[-1])
    cad_spheres.append(Sphere(r=radius, _fn=100).rotate([r[0], r[1], r[2]]).translate([translation[0], translation[1], translation[2]]))

for cone in cones:
    quaternion = Quaternion(cone[:4]) #[w,x,y,z]
    inverse = quaternion.inverse
    quaternion = np.asarray([inverse[1], inverse[2], inverse[3],inverse[0]])
    r = R.from_quat(quaternion).as_euler('xyz', degrees=True)

    translation = cone[4:7]
    angle = abs(cone[-1])
    height=100
    radius_bottom = height*angle

    cad_cones.append(Cylinder(h=height, r1=0, r2=radius_bottom, _fn=100).rotate([r[0], r[1], r[2]]).translate([translation[0], translation[1], translation[2]]))

cad_intersections = []
cad_unions = []


primitives = cad_cylinders + cad_cubes + cad_cones + cad_spheres
int_nodes = []

for i in range(cvx.shape[-1]):
    int_node = Intersection()
    for j in range(cvx.shape[0]):
        if cvx[j,i] > 0.5:
            int_node.append(primitives[j])
    int_nodes.append(int_node)

union = Union()
for i in range(ccv.shape[0]):
    if ccv[i] > 0.5:
        union.append(int_nodes[i])
union = union.scale([100, 100, 100])
union.write("sample1.scad")

print("Total Time: %f" % (time.time() - start_time))


```

**Comment by MasterOrigamii** [comment]
Thank you for your reply! I believe it is useful. So while for some reason I can't get my hands on the device and test this code, I'm closing this for now.

**Comment by bigKoki** [comment]
> 

i ran the eval.py,but it doesn't export any .npy files,i want to know what does "cvx" and "ccv" mean,thank you very much!!!

**Comment by kimren227** [comment]
That's the intersection and union connection matrices. You can find [here](https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/model.py#L155)

**Comment by bigKoki** [comment]
> That's the intersection and union connection matrices. You can find [here](https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/model.py#L155)

thanks for your reply,so i can use intersection_layer_connections, union_layer_connections and primitive_parameters in CSGStumpNet to generate a openscad file with your code in this issue?

**Comment by kimren227** [comment]
Yes, you should be able to. Let me know if you face any issues.  

Also, you may need to add an additional intersection to intersect the output with the point cloud's bounding box. This is because I forget to add padding for sample points when re-train the model for releasing the pre-train weights......

**Comment by bigKoki** [comment]
> Yes, you should be able to. Let me know if you face any issues.
> 
> Also, you may need to add an additional intersection to intersect the output with the point cloud's bounding box. This is because I forget to add padding for sample points when re-train the model for releasing the pre-train weights......

thank you,i 'll try it

**Comment by bigKoki** [comment]
> Yes, you should be able to. Let me know if you face any issues.
> 
> Also, you may need to add an additional intersection to intersect the output with the point cloud's bounding box. This is because I forget to add padding for sample points when re-train the model for releasing the pre-train weights......

thank for your code，it works，I build a function likes generate_mesh and it export valid openscad files。

**Comment by rucha97** [comment]
@kimren227 After running the eval.py code, does it generate any mesh? For me it did not, can you tell me the folder name where it gets saved?
Also, your above code takes .npy format as input. Does running the network generate any such file, I could not find any code for it.

Thanks,
Rucha

**Comment by kimren227** [comment]
> 

Yes, running eval.py should generate meshes from MachingCubes. The output meshes will be placed at "./samples/" (specified by a config file) https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/configs/bench.json#L20  If you didn't see them, you can try manually create a "./samples" folder at the project directory. 

As to generating the .npy files for exporting to OpenSCAD, you may need to edit the code a bit to save the intersection, union connection weights as well as primitive parameters. https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/model.py#L155

**Comment by rucha97** [comment]
> > 
> 
> Yes, running eval.py should generate meshes from MachingCubes. The output meshes will be placed at "./samples/" (specified by a config file)
> 
> https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/configs/bench.json#L20
> 
> If you didn't see them, you can try manually create a "./samples" folder at the project directory.
> As to generating the .npy files for exporting to OpenSCAD, you may need to edit the code a bit to save the intersection, union connection weights as well as primitive parameters.
> 
> https://github.com/kimren227/CSGStumpNet/blob/e8da70d8a9f207f9b5177c9600d153a484d15585/model.py#L155

Okay, thank you. I will try.

**Comment by rucha97** [comment]
> 

@864546664 
Hi,
Could you please share your code for generating the .scad files? I tried but facing a lot of issues with my code.
Thanks in advance!

**Comment by bigKoki** [comment]
> > 
> 
> @864546664 Hi, Could you please share your code for generating the .scad files? I tried but facing a lot of issues with my code. Thanks in advance!

you can use  intersection_layer_connections, union_layer_connections and primitive_parameters in function generate_mesh and the code above to generate a openscad file.You can have a try, if you have problem,i can give you my code here.

**Comment by rucha97** [comment]
> > > 
> > 
> > 
> > @864546664 Hi, Could you please share your code for generating the .scad files? I tried but facing a lot of issues with my code. Thanks in advance!
> 
> you can use intersection_layer_connections, union_layer_connections and primitive_parameters in function generate_mesh and the code above to generate a openscad file.You can have a try, if you have problem,i can give you my code here.

Hi, thank you for your suggestion, I tried that but something is wrong. Do you mind sharing your code snippet? TIA!

**Comment by bigKoki** [comment]
> > > > 
> > > 
> > > 
> > > @864546664 Hi, Could you please share your code for generating the .scad files? I tried but facing a lot of issues with my code. Thanks in advance!
> > 
> > 
> > you can use intersection_layer_connections, union_layer_connections and primitive_parameters in function generate_mesh and the code above to generate a openscad file.You can have a try, if you have problem,i can give you my code here.
> 
> Hi, thank you for your suggestion, I tried that but something is wrong. Do you mind sharing your code snippet? TIA!

sorry i just found that i lost the code,you can follow these steps to achieve your goal：
you can find generate_mesh in util.py，and find intersection_layer_connections, union_layer_connections and primitive_parameters。the three parameter means primitives 、cvx、ccv in the code the author offered，you can add the author's code and replace the parameter names,then you can get openscad files.

**Comment by rucha97** [comment]
@864546664 thanks for your help.

@kimren227 

`def generate_openscad(self, file_name_prefix, start_index, batch_size, intersection_layer_connections,union_layer_connections, primitive_parameters):
        Path(file_name_prefix).mkdir(parents=True, exist_ok=True)
        d1 = {'primitive': primitive_parameters, 'cvx': intersection_layer_connections,
                'ccv': union_layer_connections}
        sample_npy = np.save('sample.npy', d1)`

So, I called the above function from the generate_mesh() function in utils.py to save the primitives, intersection layers, and connection layers and  I was finally able to generate a sample.npy file.

But what I observed is that they are 3D arrays. The primitives is of shape [2,34,64] but the code you gave expects a 2D array and thus results in an error. 

Can you tell me what can I do?

Also, in model.py file I found this code
` # getting parameters for each type of primitive
boxes = primitive_parameters[:,:,:10] # [B,K/4,10]`
where B is sample_point_coordinates.shape[0].

What is sample_point_coordinates?

Please answer my queries. Thank you.

**Comment by kimren227** [comment]
> 
Sorry that I missed your message. The "B" stands for batch size. During export, you'll only need one data, for example, "intersection_layer_connections[0]".

**Comment by rucha97** [comment]
Oh, I see. Well, from the exported data I took just one data and generated the CSG file and it worked fine for sofa but didn't work for lamp. But, thank you very much for your suggestion, I will try to tweak the code more and I am sure I will get the results.
